pkgs/test_core/lib/src/runner/reporter/github.dart

```diff
@@ -0,0 +1,196 @@
+/// for a description of the output format, and
+/// https://github.com/dart-lang/test/issues/1415 for discussions about this
+/// implementation.
+class GithubReporter implements Reporter {
+  /// The engine used to run the tests.
+  final Engine _engine;
+
+  /// Whether the path to each test's suite should be printed.
+  final bool _printPath;
+
+  /// Whether the reporter is paused.
+  var _paused = false;
+
+  /// The set of all subscriptions to various streams.
+  final _subscriptions = <StreamSubscription>{};
+
+  final StringSink _sink;
+  final _helper = _GithubHelper();
+
+  final Map<LiveTest, List<Message>> _testMessages = {};
+
+  /// Watches the tests run by [engine] and prints their results as JSON.
+  static GithubReporter watch(
+    Engine engine,
+    StringSink sink, {
+    required bool printPath,
+  }) =>
+      GithubReporter._(engine, sink, printPath);
+
+  GithubReporter._(this._engine, this._sink, this._printPath) {
+    _subscriptions.add(_engine.onTestStarted.listen(_onTestStarted));
+    _subscriptions.add(_engine.success.asStream().listen(_onDone));
+  }
+
+  @override
+  void pause() {
+    if (_paused) return;
+    _paused = true;
+
+    for (var subscription in _subscriptions) {
+      subscription.pause();
+    }
+  }
+
+  @override
+  void resume() {
+    if (!_paused) return;
+    _paused = false;
+
+    for (var subscription in _subscriptions) {
+      subscription.resume();
+    }
+  }
+
+  void _cancel() {
+    for (var subscription in _subscriptions) {
+      subscription.cancel();
+    }
+    _subscriptions.clear();
+  }
+
+  /// A callback called when the engine begins running [liveTest].
+  void _onTestStarted(LiveTest liveTest) {
+    // Convert the future to a stream so that the subscription can be paused or
+    // canceled.
+    _subscriptions.add(
+        liveTest.onComplete.asStream().listen((_) => _onComplete(liveTest)));
+
+    // Collect messages from tests as they are emitted.
+    _subscriptions.add(liveTest.onMessage.listen((message) {
+      _testMessages.putIfAbsent(liveTest, () => []).add(message);
+    }));
+  }
+
+  /// A callback called when [liveTest] finishes running.
+  void _onComplete(LiveTest test) {
+    final errors = test.errors;
+    final messages = _testMessages[test] ?? [];
+    final skipped = test.state.result == Result.skipped;
+    final failed = errors.isNotEmpty;
+
+    void emitMessages(List<Message> messages) {
+      for (var message in messages) {
+        _sink.writeln(message.text);
+      }
+    }
+
+    void emitErrors(List<AsyncError> errors) {
+      for (var error in errors) {
+        _sink.writeln('${error.error}');
+        _sink.writeln(error.stackTrace.toString().trimRight());
+      }
+    }
+
+    final isLoadSuite = test.suite is LoadSuite;
+    if (isLoadSuite) {
+      // Don't emit any info for 'loadSuite' tests, unless they contain errors.
+      if (errors.isNotEmpty || messages.isNotEmpty) {
+        _sink.writeln('${test.suite.path}:');
```

Comment:
Should we also be visualizing this as a group? If not why not?

Reply:
I believe this 'LiveTest' is just plumbing used by package:test during test discovery - it's not an actual user test, and reporting it as such would be odd for the user.

We also special case these LoadSuite test in the other reporters I believe (but don't have the code in front of me right now).

Reply:
So what this is specifically is a synthetic "test" used to report the success or failure of compiling, loading, and connecting to the test. Basically, it fails if there is a compilation error in the test or some failure connecting to it.

We do want to surface it for sure, and it looks a bit weird imo to not have it show up the same way (I think it makes sense to have the actual errors hidden behind a collapsable group still for instance).

Reply:
So a `Loading test/message_extraction/failed_extraction_test.dart` group, which would generally be empty, but when it had content would be important to display (esp. as likely no other tests would run).

I wonder if we shouldn't hide it normally - as mostly it'll be empty and be noise in the rest of the output.

Reply:
If we only show it at all when it fails, then I would call it `Failed to load test/message_extraction/failed_extraction_test.dart`. But ya that SGTM

Reply:
I've updated this to show the load suite info if it has any content - messages or errors; it shows as a group, w/ a different icon from a test.